### DISCUSSION TITLE
Can't Upgrade to the latest version

### DISCUSSION THREAD

**Discussion by lexy10** [ant-media/Ant-Media-Server#6703] [Q&A]
I have been trying to update to 2.11 from 2.10 inside a docker container for over 2weeks now via ./upgrade.sh and install_ant-media-server.sh, it's not working.

Keeps giving me the error below
mv: cannot move '/usr/local/antmedia' to '/usr/local/antmedia-backup-2024-10-07_09-17-15': Device or resource busy
There is a problem in installing the ant media server. Please send the log of this console to support@antmedia.io

I have tried shutting-down with force and all, used sudo, but it's not working. Can someone help please, it's tiring and I don't want to do a fresh install, because I don't want to loose the configurations I have already. 

Or is there anyway I can do fresh install and copy some configs because I am using docker with persistent volume.

**Comment by muratugureminoglu** [comment]
Hi @lexy10 

Thanks for your question. 

In the Docker structure, you need to pull again image and since the /usr/local/antmedia directory is defined as EntryPoint, you are getting a Device or resource busy error. So the upgrade.sh command unfortunately does not work here.


My suggestion is to copy the following files to the new container. It 's also hard way.
```
/usr/local/antmedia/*.db
/usr/local/antmedia/webapps/your-apps/WEB-INF/red5-web.properties
/usr/local/antmedia/webapps/your-apps/streams
/usr/local/antmedia/conf/red5.properties
```
If you enabled the ssl; 
```
/usr/local/antmedia/conf/chain.pem,privkey.pem,fullchain.pem,truststore.jks,keystore.jks
```
However, There is the best practice. 

If you use Mongo Atlas or MongoDB and run Ant Media Server with database support, you will only need to backup the streams (/usr/local/antmedia/webapps/your-apps/streams/)directories during the upgrade process.

https://antmedia.io/docs/category/supported-databases/

Please let me know If you have further questions. 
Regards.

**Reply by lexy10** [reply]
Oh wow, thank you for the clarification. I want to move to the cluster mode now because I think that's the only way I can scale horizontally.

Meanwhile I don't save streams on the localDB, i store them in S3 and my personal DB so I don't think I need to do any backup for that.

How do you think I can move to cluster mode with the backup of my SSL and red5-web.properties?

Regards.

**Reply by muratugureminoglu** [reply]
My pleasure @lexy10 Hope it helps you.

_Oh wow, thank you for the clarification. I want to move to the cluster mode now because I think that's the only way I can scale horizontally._

If you don't want to scale it, you can as follows.

```
 ./change_server_mode.sh standalone mongodb+srv://<username>:<password>@<url>
 ```

_Meanwhile I don't save streams on the localDB, i store them in S3 and my personal DB so I don't think I need to do any backup for that._

This is good and you are right.

_How do you think I can move to cluster mode with the backup of my SSL and red5-web.properties?_

Please just copy `/usr/local/antmedia/conf/chain.pem,privkey.pem,fullchain.pem,truststore.jks,keystore.jks` files and `/usr/local/antmedia/conf/red5.properties` to your new container then run the following command.
```
 ./change_server_mode.sh standalone mongodb+srv://<username>:<password>@<url>
```

After that, your configuration will be copied to MongoDB.

Regards.

**Reply by lexy10** [reply]
Thank you for this @muratugureminoglu 

_If you don't want to scale it, you can as follows._

I actually want to scale it because we would be having 100s of stream publishers and some conferencing soon, so i want to have that settings in place already. So how does the horizontal scaling work?

I am a bit confused with it cos we have a lot of options on the docs and also especially with the server licence and the likes. Meanwhile I am using AWS Elasticbeanstalk docker environment at the moment.

So I'm confused on the right approach. Please can you help clarify? Thank you.

Regards.

**Reply by muratugureminoglu** [reply]
Hi @lexy10 

If you use autoscaling, you can choose Kubernetes or Autoscale-supported structures. Otherwise, you can use a static cluster structure. In a static cluster, scaling is only done manually, meaning you add or remove servers manually. There is no automatic Scale-up/Scale-down process. However, in Autoscale supported structures, It expands horizontally based on system load.

Kubernetes Document: https://github.com/ant-media/helm
Autoscale Cluster Document: https://antmedia.io/docs/guides/clustering-and-scaling/aws/clustering-with-aws/

You can quickly and easily set up a cluster using the CloudFormation template: https://antmedia.io/docs/guides/clustering-and-scaling/aws/scale-with-aws-cloudformation/

If you contact us at support[@]antmedia.io, we can set up a meeting and answer your questions and also help your installation.

Regards.

**Reply by lexy10** [reply]
Oh i think I understand a bit now, so what is now the use of the cluster type to standalone. What i understand it to be now is that, one server for publish and another for play. Which means i can still use standalone for publish and play and let my AWS server autoscaling group scale the servers for me based on the load right?

**Reply by muratugureminoglu** [reply]
Yes, Correct.  As I mentioned in the previous message, if you send us an e-mail tagging me, I can arrange a meeting and help you for installation. :)
